File: app/src/main/java/com/viafourasample/src/utils/InsetsUtils.java
package com.viafourasample.src.utils;

import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.widget.FrameLayout;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;
import androidx.core.graphics.Insets;
import androidx.core.view.OnApplyWindowInsetsListener;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.core.view.WindowInsetsControllerCompat;

import com.viafoura.sampleapp.R;

public final class InsetsUtils {

    private static final int SYSTEM_BARS =
            WindowInsetsCompat.Type.systemBars() | WindowInsetsCompat.Type.displayCutout();

    private InsetsUtils() {
    }

    public static void enableEdgeToEdge(AppCompatActivity activity) {
        Window window = activity.getWindow();
        WindowCompat.setDecorFitsSystemWindows(window, false);
        new WindowInsetsControllerCompat(window, window.getDecorView())
                .setAppearanceLightStatusBars(false);
    }

    public static void applyActionBarInsets(AppCompatActivity activity) {
        enableEdgeToEdge(activity);

        View decor = activity.getWindow().getDecorView();
        if (!(decor instanceof FrameLayout)) {
            return;
        }

        View scrim = decor.findViewById(R.id.status_bar_scrim);
        if (scrim == null) {
            scrim = new View(activity);
            scrim.setId(R.id.status_bar_scrim);
            scrim.setBackgroundColor(ContextCompat.getColor(activity, R.color.colorToolbar));
            ((FrameLayout) decor).addView(scrim, 0, new FrameLayout.LayoutParams(
                    FrameLayout.LayoutParams.MATCH_PARENT, 0, Gravity.TOP));
        }

        View content = decor.findViewById(android.R.id.content);
        if (content != null) {
            ViewCompat.setOnApplyWindowInsetsListener(content, new OnApplyWindowInsetsListener() {
                @Override
                public WindowInsetsCompat onApplyWindowInsets(View v, WindowInsetsCompat windowInsets) {
                    Insets bars = windowInsets.getInsets(SYSTEM_BARS);
                    v.setPadding(bars.left, bars.top, bars.right, 0);
                    return windowInsets;
                }
            });
            ViewCompat.requestApplyInsets(content);
        }

        ViewCompat.setOnApplyWindowInsetsListener(scrim, new OnApplyWindowInsetsListener() {
            @Override
            public WindowInsetsCompat onApplyWindowInsets(View v, WindowInsetsCompat windowInsets) {
                Insets bars = windowInsets.getInsets(SYSTEM_BARS);
                ViewGroup.LayoutParams params = v.getLayoutParams();
                if (params.height != bars.top) {
                    params.height = bars.top;
                    v.setLayoutParams(params);
                }
                return windowInsets;
            }
        });

        ViewCompat.requestApplyInsets(scrim);
    }
}

File: app/src/main/java/com/viafourasample/src/utils/InsetsUtils.kt
package com.viafourasample.src.utils

import android.view.Gravity
import android.view.View
import android.widget.FrameLayout
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.core.view.ViewCompat
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat
import com.viafoura.sampleapp.R

object InsetsUtils {

    private val SYSTEM_BARS =
        WindowInsetsCompat.Type.systemBars() or WindowInsetsCompat.Type.displayCutout()

    fun enableEdgeToEdge(activity: AppCompatActivity) {
        val window = activity.window
        WindowCompat.setDecorFitsSystemWindows(window, false)
        WindowInsetsControllerCompat(window, window.decorView)
            .isAppearanceLightStatusBars = false
    }

    fun applyActionBarInsets(activity: AppCompatActivity) {
        enableEdgeToEdge(activity)

        val decor = activity.window.decorView as? FrameLayout ?: return

        val existingScrim: View? = decor.findViewById(R.id.status_bar_scrim)
        val scrim = existingScrim ?: View(activity).also { view ->
            view.id = R.id.status_bar_scrim
            view.setBackgroundColor(ContextCompat.getColor(activity, R.color.colorToolbar))
            decor.addView(
                view,
                0,
                FrameLayout.LayoutParams(FrameLayout.LayoutParams.MATCH_PARENT, 0, Gravity.TOP)
            )
        }

        val content: View? = decor.findViewById(android.R.id.content)
        if (content != null) {
            ViewCompat.setOnApplyWindowInsetsListener(content) { v, windowInsets ->
                val bars = windowInsets.getInsets(SYSTEM_BARS)
                v.setPadding(bars.left, bars.top, bars.right, 0)
                windowInsets
            }
            ViewCompat.requestApplyInsets(content)
        }

        ViewCompat.setOnApplyWindowInsetsListener(scrim) { v, windowInsets ->
            val bars = windowInsets.getInsets(SYSTEM_BARS)
            val params = v.layoutParams
            if (params.height != bars.top) {
                params.height = bars.top
                v.layoutParams = params
            }
            windowInsets
        }

        ViewCompat.requestApplyInsets(scrim)
    }
}
